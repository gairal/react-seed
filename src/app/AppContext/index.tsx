import { createContext, useContext } from 'react';

import type { AppContextState } from './types';

/**
 * casting the default value to AppContextState
 * We're gonna assume that the context is always set with the required values
 */
export const AppContext = createContext<AppContextState>({} as AppContextState);

/**
 * helper hook to access the app context from any component
 */
export const useAppContext = () => useContext(AppContext);
