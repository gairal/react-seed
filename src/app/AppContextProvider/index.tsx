import React, { FC } from 'react';

import { AppContext } from './context';
import { useInitAppContext } from './useInitAppContext';

/**
 * simple main app state management using a React Context
 */
export const AppContextProvider: FC = ({ children }) => (
  <AppContext.Provider value={useInitAppContext()}>
    {children}
  </AppContext.Provider>
);
