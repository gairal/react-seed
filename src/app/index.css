.main {
  display: flex;
  flex-direction: column;
}

@screen sm {
  .main {
    flex-direction: row;
    margin: 0 -12.5px;
  }
}
