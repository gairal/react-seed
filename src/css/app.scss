/** main css file used for global styles **/

body {
  @apply p-4;
}
