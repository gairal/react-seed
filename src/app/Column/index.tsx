import React, { ChangeEvent, useState } from 'react';

import { useAppContext } from '../AppContext';
import { Col } from '../AppContext/types';
import { Task } from '../Task';

import './Column.scss';

export const Column = ({ id, title, tasks }: Col) => {
  const [taskTitle, setTaskTitle] = useState('');
  const { addTaskToColumn } = useAppContext();

  const handleChange = (event: ChangeEvent<HTMLInputElement>) =>
    setTaskTitle(event.currentTarget.value);
  const handleClick = () => {
    if (!taskTitle) {
      return;
    }
    addTaskToColumn(id, taskTitle);
    setTaskTitle('');
  };

  return (
    <section className="column">
      <header>{title}</header>
      <div>
        {tasks.map((task) => (
          <Task key={task.id} {...task} colId={id} />
        ))}
      </div>

      <form className="flex">
        <input
          className="border"
          onChange={handleChange}
          type="text"
          value={taskTitle}
        />
        <button onClick={handleClick} type="button">
          add
        </button>
      </form>
    </section>
  );
};
