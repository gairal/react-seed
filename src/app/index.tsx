import { AppContext } from './AppContext';
import { useInitAppContext } from './AppContext/useInitAppContext';
import { Header } from './Header';
import { Main } from './Main';

export const App = () => {
  return (
    <AppContext.Provider value={useInitAppContext()}>
      <Header>React Seed</Header>
      <Main />
    </AppContext.Provider>
  );
};
