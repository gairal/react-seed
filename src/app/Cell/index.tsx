import React from 'react';

import { useAppContext } from '../AppContextProvider/context';
import { Cel } from '../types';

import './Cell.scss';

interface Props extends Cel {
  colId: string;
}

export const Cell = ({ colId, id, title }: Props) => {
  const { removeItem } = useAppContext();

  const handleClick = () => removeItem(colId, id);

  return (
    <div className="cell">
      <header className="flex justify-end">
        <button onClick={handleClick} type="button">
          X
        </button>
      </header>
      <div>{title}</div>
    </div>
  );
};
