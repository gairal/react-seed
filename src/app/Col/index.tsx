import React, { ChangeEvent, useState } from 'react';

import { useAppContext } from '../AppContext';
import { Column } from '../AppContext/types';
import { Cell } from '../Cell';

import './Col.css';

export const Col = ({ id, items, title }: Column) => {
  const [taskTitle, setTaskTitle] = useState('');
  const { addTask } = useAppContext();

  const handleOnChange = ({ currentTarget }: ChangeEvent<HTMLInputElement>) => {
    setTaskTitle(currentTarget.value);
  };
  const handleClick = () => {
    addTask(id, taskTitle);
    setTaskTitle('');
  };

  return (
    <section key={id} className="col">
      <header className="col-header">{title}</header>
      <main className="col-content">
        {items.map((cell) => (
          <Cell key={cell.id} coldId={id} {...cell} />
        ))}
      </main>
      <footer className="flex justify-between items-center p-4">
        <input
          className="border"
          onChange={handleOnChange}
          type="text"
          value={taskTitle}
        />
        <button
          className="border"
          disabled={!taskTitle}
          onClick={handleClick}
          type="button"
        >
          add task
        </button>
      </footer>
    </section>
  );
};
