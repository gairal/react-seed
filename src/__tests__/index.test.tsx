import { render } from "@testing-library/react";

import { App } from "../app";

const subject = () => render(<App />);

describe("App", () => {
  it("matches snapshot", () => {
    const { asFragment } = subject();

    expect(asFragment()).toMatchSnapshot();
  });
});
