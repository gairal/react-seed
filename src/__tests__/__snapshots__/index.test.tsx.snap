// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <header>
    <h1
      class="text-3xl p-2"
    >
      React Seed
    </h1>
    <hr
      class="border mb-4"
    />
  </header>
  <main
    class="main"
  >
    <p>
      context test
    </p>
  </main>
</DocumentFragment>
`;
